net/captivedetection/endpoints.go

```diff
@@ -0,0 +1,213 @@
+		appendPlatformEndpoint("http://www.msftncsi.com/ncsi.txt", http.StatusOK, "Microsoft NCSI")
+	case "darwin", "ios":
+		appendPlatformEndpoint("http://captive.apple.com/hotspot-detect.html", http.StatusOK, "Success")
+		appendPlatformEndpoint("http://www.thinkdifferent.us/", http.StatusOK, "Success")
+		appendPlatformEndpoint("http://www.airport.us/", http.StatusOK, "Success")
+	case "android":
+		appendPlatformEndpoint("http://connectivitycheck.android.com/generate_204", http.StatusNoContent, "")
+		appendPlatformEndpoint("http://connectivitycheck.gstatic.com/generate_204", http.StatusNoContent, "")
+		appendPlatformEndpoint("http://play.googleapis.com/generate_204", http.StatusNoContent, "")
+		appendPlatformEndpoint("http://clients3.google.com/generate_204", http.StatusNoContent, "")
+	default:
+		appendPlatformEndpoint("http://detectportal.firefox.com/success.txt", http.StatusOK, "success")
+		appendPlatformEndpoint("http://network-test.debian.org/nm", http.StatusOK, "NetworkManager is online")
+	}
+
+	// Sort the endpoints by provider so that we can prioritize DERP nodes in the preferred region, followed by
+	// any other DERP server elsewhere, followed by Tailscale endpoints, and lastly any platform-specific endpoints.
+	slices.SortFunc(endpoints, func(x, y Endpoint) int {
+		return cmp.Compare(x.Provider, y.Provider)
+	})
+
+	return endpoints
+}
+
+// responseLooksLikeCaptive checks if the given HTTP response matches the expected response for the Endpoint.
+func (e Endpoint) responseLooksLikeCaptive(r *http.Response, logf logger.Logf) bool {
+	defer r.Body.Close()
+
+	// Check the status code first.
+	if r.StatusCode != e.StatusCode {
+		logf("[v1] unexpected status code in captive portal response: want=%d, got=%d", e.StatusCode, r.StatusCode)
+		return true
+	}
+
+	// If the endpoint supports the Tailscale challenge header, check that the response contains the expected header.
+	if e.SupportsTailscaleChallenge {
+		expectedResponse := "response ts_" + e.URL.Host
+		hasResponse := r.Header.Get("X-Tailscale-Response") == expectedResponse
+		if !hasResponse {
+			// The response did not contain the expected X-Tailscale-Response header, which means we are most likely
+			// behind a captive portal (somebody is tampering with the response headers).
+			logf("captive portal check response did not contain expected X-Tailscale-Response header: want=%q, got=%q", expectedResponse, r.Header.Get("X-Tailscale-Response"))
+			return true
+		}
+	}
+
+	// If we don't have an expected content string, we don't need to check the response body.
+	if e.ExpectedContent == "" {
+		return false
+	}
+
+	// Read the response body and check if it contains the expected content.
+	b, err := io.ReadAll(io.LimitReader(r.Body, 4096))
+	if err != nil {
+		logf("reading captive portal check response body failed: %v", err)
+		return false
+	}
+	hasExpectedContent := strings.Contains(string(b), e.ExpectedContent)
```

Comment:
nit: you can use https://pkg.go.dev/go4.org/mem#Contains to avoid allocating, not that this will be super hot

mem.Contains(mem.B(b), mem.S(e.ExpectedContent))